### PR TITLE
[FLINK-4552] Refactor WindowOperator/Trigger Tests

### PR DESCRIPTION
**Pull request by aljoscha**
This builds on #2570 

Before, tests for WindowOperator, WindowAssigner, Trigger and WindowFunction
were all conflated in WindowOperatorTest. All of these tested that a
certain combination of a Trigger, WindowAssigner and WindowFunction produce
the expected output.

This change modularizes these tests and spreads them out across multiple
files. For example, one per trigger/window assigner.

The new WindowOperatorTest tests verify that the interaction between
WindowOperator and the various other parts works as expected, that the
correct methods on Trigger and WindowFunction are called at the expected
time and that snapshotting, timers, cleanup etc. work correctly. These tests
also verify that the different state types and WindowFunctions work correctly.

For trigger tests this introduces TriggerTestHarness. This can be used
to inject elements into Triggers they fire at the correct times. The
actual output of the WindowFunction is not important for these tests.
The new tests also make sure that triggers correctly clean up state and timers.

WindowAssigner tests verify the behaviour of window assigners in isolation.
They also test, for example, whether offset parameter of time-based windows
work correctly.

R: @StephanEwen @StefanRRichter @kl0u for review

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @StefanRRichter @kl0u , this sits on top of #2570 so you can review that one first and then this one additional commit here.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.FIRE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(2, 4)));
+
+		// right now, CountTrigger will clear it's state in onElement when firing
+		// ideally, this should be moved to onFire()
+		assertEquals(1, testHarness.numStateEntries());
+		assertEquals(0, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+	}
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java]: You could also let W(2,4) fire for completeness, so that the count was not reset by the previous firing.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(2, 4)));
+
+		// shouldn't have any timers
+		assertEquals(0, testHarness.numProcessingTimeTimers());
+		assertEquals(0, testHarness.numEventTimeTimers());
+
+		assertEquals(2, testHarness.numStateEntries());
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java]: You could test this a little bit more, e.g. with a wider window that also subsumes both, or 3 windows and just merging two, etc. to catch corner cases

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeSessionWindowsWindowsTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+ * Tests for {@link EventTimeSessionWindows}
+ */
+public class EventTimeSessionWindowsWindowsTest extends TestLogger {
+
+	@Test
+	public void testWindowAssignment() {
+		WindowAssigner.WindowAssignerContext mockContext =
+				mock(WindowAssigner.WindowAssignerContext.class);
+
+		EventTimeSessionWindows assigner = EventTimeSessionWindows.withGap(Time.milliseconds(5000));
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeSessionWindowsWindowsTest.java]: Very minor, but I think a named constant for the value 5000 might make the contracts even clearer (e.g. int gap = 5000) and then do (+ gap) to window start time to obtain end time.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeTriggerTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		assertEquals(TriggerResult.FIRE, testHarness.advanceWatermark(2, new TimeWindow(0, 2)));
+
+		assertEquals(0, testHarness.numStateEntries());
+		assertEquals(0, testHarness.numProcessingTimeTimers());
+		assertEquals(1, testHarness.numEventTimeTimers());
+		assertEquals(0, testHarness.numEventTimeTimers(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numEventTimeTimers(new TimeWindow(2, 4)));
+
+		assertEquals(TriggerResult.FIRE, testHarness.advanceWatermark(4, new TimeWindow(2, 4)));
+
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeTriggerTest.java]: We could also check that multiple trigger fire when a watermark surpassed them all at once (and maybe not all if we use 3 trigger for this), to catch corner cases.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/PurgingTriggerTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * Tests for {@link PurgingTrigger}.
+ */
+public class PurgingTriggerTest {
+
+	/**
+	 * Check if {@link PurgingTrigger} implements all methods of {@link Trigger}, as a sanity
+	 * check.
+	 */
+	@Test
+	public void testAllMethodsImplemented() throws NoSuchMethodException {
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/PurgingTriggerTest.java]: What is the purpose of this and how is it different from checking that PurgingTrigger is not abstract?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+	 * is returned for that firing.
+	 */
+	public TriggerResult advanceProcessingTime(long time, W window) throws Exception {
+		Collection<Tuple2<W, TriggerResult>> firings = advanceProcessingTime(time);
+
+		if (firings.size() != 1) {
+			throw new IllegalStateException("Must have exactly one timer firing. Fired timers: " + firings);
+		}
+
+		Tuple2<W, TriggerResult> firing = firings.iterator().next();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java]: I don't think this checks EXACTLY one, but only AT LEAST one.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+	 * is returned for that firing.
+	 */
+	public TriggerResult advanceWatermark(long time, W window) throws Exception {
+		Collection<Tuple2<W, TriggerResult>> firings = advanceWatermark(time);
+
+		if (firings.size() != 1) {
+			throw new IllegalStateException("Must have exactly one timer firing. Fired timers: " + firings);
+		}
+
+		Tuple2<W, TriggerResult> firing = firings.iterator().next();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java]: Same here.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java (hunk 1 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java]: Done

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java (hunk 2 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java]: Done, theres are also more tests about merging specifically in `MergingWindowSetTest` and `WindowOperatorTest`. The core of the merging "algorithm" is exercised in `MergingWindowSetTest`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeSessionWindowsWindowsTest.java (hunk 3 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeSessionWindowsWindowsTest.java]: Done

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeTriggerTest.java (hunk 4 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EventTimeTriggerTest.java]: The trigger tests only test the isolated behaviour of Triggers, there are no multiple keys here so that wouldn't work. You're right though, about the multiple timers so I'll add a test for this to `WindowOperatorTest` where the timer firing behaviour is tested.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/PurgingTriggerTest.java (hunk 5 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/PurgingTriggerTest.java]: `Trigger` has `canMerge()` and `onMerge()` which are not abstract in `Trigger` (`clear()` used to not be abstract but is now, after we recently changed that on master). The purging trigger can therefore be non-abstract and still not implement these methods, therefore the checks.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java (hunk 6 above)
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java]: With the check in the lines above (that checks for the size of `firings`) it should be correct, right?

**Comment by aljoscha** [comment]
Thanks for your review, @StefanRRichter! You found some areas where this can be improved.

I'm not yet sure whether parameterisation for processing-time/event-time can be done without obscuring what the test actually does to much but I'll try and come up with something.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java (hunk 6 above)
Comment (StefanRRichter) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java]: True :-)

**Comment by StefanRRichter** [comment]
My idea would be to introduce an adapter that abstracts advancing of time, where two concrete implementations exist that forward to advanceWatermark or on onProcessingTime. Or am I missing something that fundamentally more different?

**Comment by aljoscha** [comment]
No, it's more or less that but there is more stuff that needs to be in that adapter, for example in `testOnProcessingTimeFire()`, I'm highlighting the places that need changing:

```
public void testOnProcessingTimeFire() throws Exception {

    WindowAssigner<Integer, TimeWindow> mockAssigner = mockTimeWindowAssigner();
    when(mockAssigner.isEventTime()).thenReturn(false); <-- here
    Trigger<Integer, TimeWindow> mockTrigger = mockTrigger();

    KeyedOneInputStreamOperatorTestHarness<Integer, Integer, WindowedValue<List<Integer>, TimeWindow>> testHarness =
            createListWindowOperator(mockAssigner, mockTrigger, 0L);

    testHarness.open();

    testHarness.setProcessingTime(Long.MIN_VALUE); <-- here

    when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
            .thenReturn(Arrays.asList(new TimeWindow(2, 4), new TimeWindow(0, 2)));

    assertEquals(0, testHarness.getOutput().size());
    assertEquals(0, testHarness.numKeyedStateEntries());

    doAnswer(new Answer<TriggerResult>() {
        @Override
        public TriggerResult answer(InvocationOnMock invocation) throws Exception {
            Trigger.TriggerContext context = (Trigger.TriggerContext) invocation.getArguments()[3];
            context.registerProcessingTimeTimer(0L); <-- here
            context.getPartitionedState(valueStateDescriptor).update("hello");
            return TriggerResult.CONTINUE;
        }
    }).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());

    shouldFireOnProcessingTime(mockTrigger); <-- here

    testHarness.processElement(new StreamRecord<>(0, 0L));

    assertEquals(4, testHarness.numKeyedStateEntries()); // window-contents and trigger state for two windows
    assertEquals(4, testHarness.numProcessingTimeTimers()); // timers/gc timers for two windows <-- here

    testHarness.setProcessingTime(0L); <-- here

    // clear is only called at cleanup time/GC time
    verify(mockTrigger, never()).clear(anyTimeWindow(), anyTriggerContext());

    // FIRE should not purge contents
    assertEquals(4, testHarness.numKeyedStateEntries());
    assertEquals(2, testHarness.numProcessingTimeTimers()); // only gc timers left <-- here

    // there should be two elements now
    assertThat(testHarness.extractOutputStreamRecords(),
            containsInAnyOrder(
                    isWindowedValue(contains(0), 1L, timeWindow(0, 2)),
                    isWindowedValue(contains(0), 3L, timeWindow(2, 4))));
}
```

(man I can't make text bold in a code block ... 😭)

**Comment by StefanRRichter** [comment]
Yes, I was aware that some configuration/mocking also depends on the time domain. However what those calls do is essentially equivalent, just for different time domain. So I assume an adapter that just dispatches for different time domains should not obscure the logic of the tests?

**Comment by aljoscha** [comment]
Yeah, alright, I'm just lazy. 😅 I'll whip something up.

**Comment by StefanRRichter** [comment]
Yeah, I know the pain ;-) but next times when we need to fix only half the amount of tests on a change it will quickly pay off (hopefully)

**Comment by aljoscha** [comment]
@StefanRRichter I added more tests (see the commit) and deduplicated processing-time/event-time tests. PTAL (please take another look) 😃

**Comment by StefanRRichter** [comment]
Had another look at the changes, I think it is a very well written test! +1 for merge.

**Comment by aljoscha** [comment]
manually merged
